### PR TITLE
Finally remove boost!

### PR DESCRIPTION
**Pull request by rcurtin**
This depends on #3208 being merged first.  But, after that is merged, there are no uses of Boost in mlpack (excepting all the stuff we haven't yet adapted after #2777).  So, we can remove it!

This PR removes the dependency and updates documentation accordingly.

### CODE REVIEW

File: doc/guide/build.hpp [hunk 1]
```diff
@@ -140,8 +138,8 @@ source as apt installs an older version. So you need to omit
 On Fedora, Red Hat, or CentOS, these same dependencies can be obtained via dnf:
 
 @code
-# dnf install boost-devel boost-math armadillo-devel binutils-devel
-  python3-Cython python3-setuptools python3-numpy python3-pandas ensmallen-devel
+# dnf install armadillo-devel binutils-devel python3-Cython python3-setuptools
+    python3-numpy python3-pandas ensmallen-devel
```
Comment (zoq) [inline doc/guide/build.hpp:142]: Is the extra space on purpose?

Comment (rcurtin) [reply doc/guide/build.hpp]: Nope, I added it accidentally.  Fixed in 9910d24. :+1:

**Review by shrit — APPROVED** [review]
LGTM. Failing tests are not related.
We have to handle them at some point.

**Review by zoq — APPROVED** [review]
Nice 👍

**Pull request merged**